In [27]:
from scipy.sparse import csr_matrix
import numpy as np

In [28]:
X = np.asarray([[0, 1], [2, 3]])
y = np.asarray([1, 2])
print X.shape, y.shape
print "X", X
print "y", y
print X.dot(y)
print y.dot(X).shape

(2, 2) (2,)
X [[0 1]
 [2 3]]
y [1 2]
[2 8]
(2,)


In [71]:
# Here we are trying to find the fastest way to sum sparse matrix over rows
# There are two possibilities:
# built-in sum method
# left multiply X by row-vector filled with ones

D = 100
data = range(D)
row = range(D)
col = range(D)
X = csr_matrix((data, (row, col)), shape=(D, D))
y = np.ones(D)
print "Use sum method"
%timeit X.sum(axis=0)
%timeit X.sum(axis=1)

print 20 * "="
print "Use dot product"
#%timeit y.dot(X)       # This is a wrong way in case of sparse matrices
%timeit(csr_matrix(np.ones(D)).dot(X))
%timeit X.dot(np.ones(D))

Use sum method
10000 loops, best of 3: 90.7 µs per loop
10000 loops, best of 3: 27.3 µs per loop
Use dot product
1000 loops, best of 3: 257 µs per loop
The slowest run took 4.75 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 3: 6.33 µs per loop


In [72]:
# Now, check the complexity as the dimensionality grows
D = 1000
data = range(D)
row = range(D)
col = range(D)
X = csr_matrix((data, (row, col)), shape=(D, D))
print "Use sum method"
%timeit X.sum(axis=0)
%timeit X.sum(axis=1)

print 20 * "="
print "Use dot product"
#%timeit np.ones(D).dot(X)
%timeit(csr_matrix(np.ones(D)).dot(X))
%timeit X.dot(np.ones(D))

Use sum method
10000 loops, best of 3: 99.2 µs per loop
The slowest run took 5.33 times longer than the fastest. This could mean that an intermediate result is being cached.
10000 loops, best of 3: 45.6 µs per loop
Use dot product
1000 loops, best of 3: 304 µs per loop
The slowest run took 6.22 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 3: 11 µs per loop


In [74]:
# Now, check the complexity as the dimensionality grows
D = 20000
data = range(D)
row = range(D)
col = range(D)
X = csr_matrix((data, (row, col)), shape=(D, D))
print "Use sum method"
%timeit X.sum(axis=0)
%timeit X.sum(axis=1)

print 20 * "="
print "Use dot product"
#%timeit np.ones(D).dot(X)
%timeit(csr_matrix(np.ones(D)).dot(X))
%timeit X.dot(np.ones(D))

Use sum method
10000 loops, best of 3: 168 µs per loop
1000 loops, best of 3: 341 µs per loop
Use dot product
1000 loops, best of 3: 1.06 ms per loop
10000 loops, best of 3: 81.8 µs per loop


In [69]:
# Now check types and shapes of results
a = X.sum(axis=0)
b = X.sum(axis=1)
c = y.dot(X)
c2 = csr_matrix(np.ones(D)).dot(X)
d = X.dot(y)
print type(a), type(b), type(c), type(c)
print c[0:2]  #  oops
print c2[0, 1]

100
<class 'numpy.matrixlib.defmatrix.matrix'> <class 'numpy.matrixlib.defmatrix.matrix'> <type 'numpy.ndarray'> <type 'numpy.ndarray'>
[ <100x100 sparse matrix of type '<type 'numpy.float64'>'
	with 100 stored elements in Compressed Sparse Row format>
 <100x100 sparse matrix of type '<type 'numpy.float64'>'
	with 100 stored elements in Compressed Sparse Row format>]
1.0


In [80]:
# Let's try non-diagonal X
print 20 * "="
print "D = 1000"
D = 1000
data = range(2 * D - 1)
row = [0] * D + range(1, D)
col = range(D) + [0] * (D - 1)
X = csr_matrix((data, (row, col)), shape=(D, D))
print "Use sum method"
%timeit X.sum(axis=0)
print 20 * "="
print "Use dot product"
#%timeit np.ones(D).dot(X)
%timeit(csr_matrix(np.ones(D)).dot(X))


# Let's try non-diagonal X
print 20 * "="
print "D = 10000"
D = 10000
data = range(2 * D - 1)
row = [0] * D + range(1, D)
col = range(D) + [0] * (D - 1)
X = csr_matrix((data, (row, col)), shape=(D, D))
print "Use sum method"
%timeit X.sum(axis=0)
print 20 * "="
print "Use dot product"
#%timeit np.ones(D).dot(X)
%timeit(csr_matrix(np.ones(D)).dot(X))


D = 1000
Use sum method
10000 loops, best of 3: 98.7 µs per loop
Use dot product
1000 loops, best of 3: 306 µs per loop
D = 10000
Use sum method
10000 loops, best of 3: 149 µs per loop
Use dot product
1000 loops, best of 3: 724 µs per loop


In [88]:
# Now, finally let's try pure ndarray
print 20 * "="
print "D = 100"
D = 100

X = np.zeros((D, D))
X[0, :] = range(D)
X[:, 0] = range(D, 2 * D)
%timeit X.sum(axis=0)
print 20 * "="
print "Use dot product"
%timeit np.ones(D).dot(X)

print 20 * "="
print "D = 1000"
D = 1000

X = np.zeros((D, D))
X[0, :] = range(D)
X[:, 0] = range(D, 2 * D)
%timeit X.sum(axis=0)
print 20 * "="
print "Use dot product"
%timeit np.ones(D).dot(X)


D = 100
The slowest run took 4.82 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 3: 7.66 µs per loop
Use dot product
The slowest run took 7.07 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 3: 16.3 µs per loop
D = 1000
1000 loops, best of 3: 589 µs per loop
Use dot product
100 loops, best of 3: 2.02 ms per loop


Results for reducing the first dimension

Diagonal matrix

| D  | 100 | 1000 | 20000 |
|-- |--|--|--|
| sum  | 91$\mu$s | 99$\mu$s | 168$\mu$s |
| dot  | 257$\mu$s | 300$\mu$s | 1000$\mu$s |

Non-diagonal matrix

| D  | 1000 | 20000 |
|-- |--|--|--|
| sum  | 99$\mu$s | 149$\mu$s |
| dot  | 306$\mu$s | 724$mu$s |

Sublinear growth of time in number of elements (we were afraid it will be D^2)???
Seems like lot's of bookkeeping is going on here and constant is huge?


Pure ndarray + Non-diagonal matrix

| D  | 100 | 1000 |
|-- |--|--|--|
| sum  | 8$\mu$s | 590$\mu$s |
| dot  | 16$\mu$s | 2000$mu$s |

Yay! roughly 100x slowdown, as expected from D^2 growth of time


We conclude, that plain sum(axis=0) is efficient for sum-collapsing the rows of csr matrix.
Seems like it efficiently utilizes sparsity of the data. The same is true for sum 